### PR TITLE
Cover MSDataStorage with unit tests

### PR DESCRIPTION
**Pull request by dtrabo**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

This PR covers `MSDataStorage` class with unit tests:
- Test enabled state;
- Test create golden path;
- Test create but token exchange failed;
- Test create but bad HTTP code returned;
- Test create but response deserialization failed;
- Test delete golden path;
- Test delete but token exchange failed;
- Test delete but bad HTTP code returned.

During testing a bug with deserialization was found so in addition this PR:
- Fixes deserialization error handling (6f1444151a200abaa27d90b4950b80c0fa0fdeb7).

## Misc

- Tests coverage:
Screen Shot 2019-03-28 at 4 39 05 PM
- Some methods like `list*` will be covered in another PRs (#1436).
- `writeOptions` argument is unused so method overloads with write options are left uncovered.

---

[AB#58879](https://msmobilecenter.visualstudio.com/web/wi.aspx?pcguid=27bf89de-9697-418a-accf-1aa125b22914&id=58879)

### CODE REVIEW

**Comment by jwallra** [comment]
@dtrabo @MatkovIvan Please check the `feature/storage-list-documents` branch - the DataStore tests were set up very haphazardly and a lot of changes had to be made (re-targeting, re-linking, etc).  It would be good to merge those changes into your branch.  We will be covering the new "list" methods added, so for this PR just focus on the other methods.

**Comment by roland-chernov-akvelon** [comment]
@jwallra @yochaude FYI: There were changes made in `develop` so I changed the base branch of this PR from `feature/storage-list-documents` to `develop`. Unfortunately, these tests won't be working for #1436 since CosmodDB completion handler arguments signature was changed but a fix should be simple: fdf6246fffc70a4509e9335769ffc77f1ef9e8ab.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 1]
```diff
@@ -378,12 +380,19 @@ - (void)performOperationForPartition:(NSString *)partition
...
-      sharedInstance = [self new];
+      sharedInstance = [[MSDataStore alloc] init];
     }
   });
   return sharedInstance;
 }
 
++ (void)resetSharedInstance {
+
+  // resets the once_token so dispatch_once will run again.
```
Comment (MatkovIvan) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]:
```suggestion
  // Resets the once_token so dispatch_once will run again.
```

File: AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m [hunk 2]
```diff
@@ -180,4 +257,315 @@ - (void)testPerformCosmosDbAsyncOperationWithHttpClient {
...
+  NSString *expectedDocumentId = @"testDocumentId";
+  NSString *expectedPartition = @"testPartition";
+  id<MSSerializableDocument> mockSerializableDocument = [MSFakeSerializableDocument new];
+  __block BOOL completionHandlerCalled = NO;
+  __block MSDocumentWrapper *actualDocumentWrapper;
+
+  // Mock tokens fetching.
+  MSTokenResult *testToken = [[MSTokenResult alloc] initWithDictionary:[self prepareMutableDictionary]];
+  MSTokensResponse *testTokensResponse = [[MSTokensResponse alloc] initWithTokens:@[ testToken ]];
+  OCMStub([self.tokenExchangeMock performDbTokenAsyncOperationWithHttpClient:OCMOCK_ANY partition:OCMOCK_ANY completionHandler:OCMOCK_ANY])
```
Comment (clpolet) [inline AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m]: DataStore is supposed to call token exchange client with the right partition. Let's make sure of that.
```suggestion
  OCMStub([self.tokenExchangeMock performDbTokenAsyncOperationWithHttpClient:OCMOCK_ANY partition: expectedPartition completionHandler:OCMOCK_ANY])
```

Comment (roland-chernov-akvelon) [reply AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m]: Yeah, you're right. Fixed. Thanks.

**Pull request merged**